waffle-chai/src/matchers/emit.ts

```diff
@@ -1,70 +1,88 @@
-  const filterLogsWithTopics = (logs: providers.Log[], topic: any, contractAddress: string) =>
+  const filterLogsWithTopics = (logs: providers.Log[], topic: any, contractAddress?: string) =>
     logs.filter((log) => log.topics.includes(topic))
-      .filter((log) => log.address && log.address.toLowerCase() === contractAddress.toLowerCase());
+      .filter((log) =>
+        log.address &&
+        (contractAddress === undefined || log.address.toLowerCase() === contractAddress.toLowerCase()
+        ));
 
-  Assertion.addMethod('emit', function (this: any, contract: Contract, eventName: string) {
+  const assertEmit = (assertion: any, frag: utils.EventFragment, isNegated: boolean, from?: string) => {
+    const topic = keccak256(toUtf8Bytes(frag.format()));
+    const receipt: providers.TransactionReceipt = assertion.txReceipt;
+    assertion.args = filterLogsWithTopics(receipt.logs, topic, from);
+    const isCurrentlyNegated = assertion.__flags.negate === true;
+    assertion.__flags.negate = isNegated;
+    assertion.assert(assertion.args.length > 0,
+      `Expected event "${frag.name}" to be emitted, but it wasn't`,
+      `Expected event "${frag.name}" NOT to be emitted, but it was`
+    );
+    assertion.__flags.negate = isCurrentlyNegated;
+  };
+
+  Assertion.addMethod('emit', function (this: any, contractOrEventSig: Contract|string, eventName?: string) {
     if (typeof this._obj === 'string') {
+      if (typeof contractOrEventSig === 'string') {
+        throw new Error('The emit by event signature matcher must be called on a transaction');
+      }
       // Handle specific case of using transaction hash to specify transaction. Done for backwards compatibility.
-      this.callPromise = waitForPendingTransaction(this._obj, contract.provider)
+      this.callPromise = waitForPendingTransaction(this._obj, contractOrEventSig.provider)
         .then(txReceipt => {
           this.txReceipt = txReceipt;
         });
     } else {
       callPromise(this);
     }
     const isNegated = this.__flags.negate === true;
-    this.callPromise = this.callPromise
-      .then(() => {
-        if (!('txReceipt' in this)) {
-          throw new Error('The emit matcher must be called on a transaction');
+    this.callPromise = this.callPromise.then(() => {
+      if (!('txReceipt' in this)) {
+        throw new Error('The emit matcher must be called on a transaction');
+      }
+      let eventFragment: utils.EventFragment | undefined;
+      if (typeof contractOrEventSig === 'string') {
+        try {
+          eventFragment = utils.EventFragment.from(contractOrEventSig);
+        } catch (e) {
+          throw new Error(`Invalid event signature: "${contractOrEventSig}"`);
         }
-        const receipt: providers.TransactionReceipt = this.txReceipt;
-        let eventFragment: utils.EventFragment | undefined;
+        assertEmit(this, eventFragment, isNegated);
+      } else if (typeof contractOrEventSig === 'object' && eventName) {
```

Comment:
`typeof contractOrEventSig === 'object'` - is it really necessary? I think typescript type system should do this for us